luda-editor/luda-editor-client/src/app/editor/history.rs

```diff
@@ -0,0 +1,109 @@
+pub struct History<TState> {
+    undo_count: usize,
+    initial_state: TState,
+    states: Vec<TState>,
+}
+
+impl<TState> History<TState> {
+    pub fn new(initial_state: TState) -> Self {
+        Self {
+            undo_count: 0,
+            initial_state,
+            states: Vec::new(),
+        }
+    }
+    pub fn push(&mut self, state: TState) {
+        for _ in 0..(self.undo_count.min(self.states.len())) {
+            self.states.pop();
+        }
+        self.undo_count = 0;
+        self.states.push(state);
+    }
+    pub fn undo(&mut self) -> Option<()> {
+        if self.undo_count < self.states.len() {
+            self.undo_count += 1;
+            Some(())
+        } else {
+            None
+        }
+    }
```

Comment:
Isn't it appropriate to use Result rather than Option?

Reply:
Result means OK or Err, I think it's not a error if there are nothing to undo.